### ISSUE TITLE
MemorySanitizer: Use-of-uninitialized-value

### ISSUE THREAD

**Issue by ghost** [madler/zlib#512]
Hello Mark,

I have a project that runs on Google OSS-Fuzz where I'm getting the following msan report for a specific configuration:

```
==269==WARNING: MemorySanitizer: use-of-uninitialized-value
  | #0 0x7f9bb9072e6a in longest_match /work/zlib-1.2.8.dfsg/deflate.c:1228:58
  | #1 0x7f9bb9066d20 in deflate_fast /work/zlib-1.2.8.dfsg/deflate.c:1665:31
  | #2 0x7f9bb90561b0 in deflate /work/zlib-1.2.8.dfsg/deflate.c:905:25
```

[...]

```
Uninitialized value was stored to memory at
  | #0 0x4cb326 in __msan_memcpy /src/llvm-project/compiler-rt/lib/msan/msan_interceptors.cpp:1558:3
  | #1 0x7f9bb9042b8a in fill_window /work/zlib-1.2.8.dfsg/deflate.c:1421:13
  | #2 0x7f9bb9066405 in deflate_fast /work/zlib-1.2.8.dfsg/deflate.c:1642:13
  | #3 0x7f9bb90561b0 in deflate /work/zlib-1.2.8.dfsg/deflate.c:905:25
```

This only happens (from what I have seen) with a specific deflateInit2:

```
deflateInit2(&deflationStream, 1, Z_DEFLATED, -9, 1, Z_DEFAULT_STRATEGY);
```

and with deflate call:

```
deflate(&deflationStream, Z_SYNC_FLUSH);
```

where before deflateInit2:

```
z_stream deflationStream = {};
```

It does not happen with other memLevel or windowBit values (at least I haven't seen any other conf. that triggers this report).

Is this case fuzzed on OSS-Fuzz? Because I'm assuming Zlib is on OSS-Fuzz?

**Comment by zmodem** [comment]
I don't know about OSS-Fuzz, but Chromium's fuzzer hits this, and we carry a patch to initialize the window: https://chromium-review.googlesource.com/c/chromium/src/+/2517520

We didn't have a reproducer for the regular fill_window function though, only an SSE version which doesn't exist upstream.

**Comment by madler** [comment]
Can you (@ghost) provide code to reproduce this?

**Comment by Neustradamus** [comment]
@madler: Ghost is the general GitHub account for users who have removed this account badly...

**Comment by madler** [comment]
Ah. I guess we'll have to wait for this to pop up again from someone else. Closing.

**Comment by Neustradamus** [comment]
@zmodem, @Adenilson, @ProgramMax: Can you look here?
Thanks in advance.

**Comment by zmodem** [comment]
I haven't been able to reproduce the original report. As mentioned above, we've hit the same MSan error, but with a different version of `fill_window`. Our bugs for this were:
https://bugs.chromium.org/p/chromium/issues/detail?id=1137613
https://bugs.chromium.org/p/chromium/issues/detail?id=1144420

I've run our fuzzer a bit on vanilla zlib 1.2.8 (what the OP seems to have used) today, but did not hit this.

Waiting to see if it pops up again from someone else sounds reasonable.

**Comment by Neustradamus** [comment]
@madler: Have you seen the @zmodem answer?

But @zmodem: Zlib 1.2.13 (2022-10-12) has been released and 1.2.8 (2013-04-29), more 10 years!

**Comment by madler** [comment]
I am not authorized to access the reproducer test cases.

**Comment by zmodem** [comment]
> I am not authorized to access the reproducer test cases.

They're also attached to the bug entries.
https://bugs.chromium.org/p/chromium/issues/detail?id=1137613#c9
https://bugs.chromium.org/p/chromium/issues/detail?id=1144420#c7

But they're probably not very useful as I don't recall being able to reproduce these with upstream zlib.

> But @zmodem: Zlib 1.2.13 (2022-10-12) has been released and 1.2.8 (2013-04-29), more 10 years!

Yes, the point is that I'm not able to reproduce the issue the OP was seeing with the version they presumably used.

**Comment by madler** [comment]
I don't know what you mean by "attached", but the Reproducer Testcase links on those two pages are what I'm saying I am not authorized to access. In any case, I will just assume that your inability to reproduce the issue supports the closure of this issue.
